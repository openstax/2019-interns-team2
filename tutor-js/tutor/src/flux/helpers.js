import _ from 'underscore';
import flux from 'flux-react';

const LOADING = 'loading';
const LOADED  = 'loaded';
const FAILED  = 'failed';
const SAVING  = 'saving';
const DELETING = 'deleting';
const DELETED = 'deleted';

const STATES = {
  LOADING,
  LOADED,
  FAILED,
  SAVING,
  DELETING,
  DELETED,
};

let idCounter = 0;
const CREATE_KEY = () => `_CREATING_${idCounter++}`;

const isNew = id => /_CREATING_/.test(id);

const CrudConfig = function() {
  return {
    _asyncStatus: {},
    _local: {},
    _changed: {},
    _errors: {},
    _reload: {},

    // If the specific type needs to do something else to the object:
    // _loaded : (obj, id) ->
    // _saved : (obj, id) ->
    // _reset : ->

    reset() {
      this._asyncStatus = {};
      this._local = {};
      this._changed = {};
      this._errors = {};
      this._reload = {};

      if (typeof this._reset === 'function') {
        this._reset();
      }
      return this.emitChange();
    },

    dontReload(id) {
      return (this._asyncStatus[id] === LOADED) && this._HACK_DO_NOT_RELOAD;
    },

    FAILED(status, msg, id, ...args) {
      this._asyncStatus[id] = FAILED;
      this._errors[id] = msg;
      if (typeof this._failed === 'function') {
        this._failed({ status, msg }, id, ...Array.from(args));
      }

      if (status !== 0) { // indicates network failure
        delete this._local[id];
        return this.emitChange();
      }
    },

    load(id) {
      return this._load(id);
    },

    // allow aliasing load without triggering api call.
    _load(id) {
      // Add a shortcut for unit testing
      if (this.dontReload(id)) { return; }
      this._reload[id] = false;
      this._asyncStatus[id] = LOADING;
      return this.emitChange();
    },

    loaded(obj, id, ...args) {
      // id = obj.id
      this._asyncStatus[id] = LOADED;

      if (obj) {
        // If the specific type needs to do something else to the object:
        this._local[id] = (typeof this._loaded === 'function' ? this._loaded(obj, id, ...Array.from(args)) : undefined) || obj;
      }

      return this.emitChange();
    },

    save(id, obj) {
      // Note: id could be isNew()
      this._asyncStatus[id] = SAVING;
      return this.emitChange();
    },

    saved(result, id, ...args) {
      // id = result.id
      this._asyncStatus[id] = LOADED; // TODO: Maybe make this SAVED

      // If the specific type needs to do something else to the object:
      const obj = typeof this._saved === 'function' ? this._saved(result, id, ...Array.from(args)) : undefined;
      if (obj) { result = obj; }

      if (result) {
        this._local[id] = result;
        this._local[result.id] = result;
        delete this._changed[result.id];
      } else {
        console.warn('API WARN: Server did not return JSON after saving. Patching locally');
        // Merge all the local changes into the new local object
        this._local[id] = _.extend(this._local[id], this._changed[id]);
      }
      delete this._changed[id];
      delete this._errors[id];
      // If the specific type needs to do something else to the object:
      return this.emitChange();
    },

    create(localId, attributes = {}) {
      if (!isNew(localId)) { throw new Error('BUG: MUST provide a local id'); }
      this._local[localId] = {};
      this._changed[localId] = attributes;
      return this._asyncStatus[localId] = LOADED;
    },

    created(result, localId, ...args) {
      this._local[localId] = result; // HACK: So react component can still manipulate the same object
      this._local[result.id] = result;
      this._asyncStatus[localId] = LOADED;
      this._asyncStatus[result.id] = LOADED;
      this.emitChange();
      return (typeof this._created === 'function' ? this._created(result, localId, ...Array.from(args)) : undefined);
    },

    _change(id, obj) {
      if (this._changed[id] == null) { this._changed[id] = {}; }
      _.extend(this._changed[id], obj);
      return this.emitChange();
    },

    _save(id) {
      return this._asyncStatus[id] = SAVING;
    },

    delete(id) {
      this._asyncStatus[id] = DELETING;
      return (typeof this._delete === 'function' ? this._delete(id) : undefined);
    },

    deleted(result, id, ...args) {
      this._asyncStatus[id] = DELETED;
      delete this._local[id];
      this.emitChange();
      return (typeof this._deleted === 'function' ? this._deleted(result, id, ...Array.from(args)) : undefined);
    },

    clearChanged(id) {
      return delete this._changed[id];
    },

    HACK_DO_NOT_RELOAD(bool) { return this._HACK_DO_NOT_RELOAD = bool; },

    // Keep this here so other exports method have access to it
    _get(id) {
      const val = this._local[id];
      if (!val && (this._asyncStatus[id] !== SAVING)) { return null; }
      // Scores stores an Array unlike most other stores
      if (_.isArray(val)) {
        return val;
      } else {
        return _.extend({}, val, this._changed[id]);
      }
    },

    exports: {
      isUnknown(id) { return !this._asyncStatus[id]; },
      isLoading(id) { return this._asyncStatus[id] === LOADING; },
      isLoaded(id) { return this._asyncStatus[id] === LOADED; },
      isDeleting(id) { return this._asyncStatus[id] === DELETING; },
      isSaving(id) { return this._asyncStatus[id] === SAVING; },
      isFailed(id) { return this._asyncStatus[id] === FAILED; },
      getAsyncStatus(id) { return this._asyncStatus[id]; },
      get(id) { return this._get(id); },
      isChanged(id) { return !_.isEmpty(this._changed[id]); },
      getChanged(id) { return this._changed[id] || {}; },
      freshLocalId() { return CREATE_KEY(); },
      isNew(id) { return isNew(id); },
      reload(id) { return this._reload[id]; },
      getError(id) { return this._errors[id]; },
    },
  };
};

// Helper for creating a simple store for actions
const makeSimpleStore = function(storeConfig) {

  let actionsConfig = _.omit(storeConfig, (value, key) => (typeof value !== 'function') || (key === 'exports'));

  actionsConfig = _.keys(actionsConfig);
  const actions = flux.createActions(actionsConfig);
  storeConfig.actions = _.values(actions);
  const store = flux.createStore(storeConfig);
  return { actions, store };
};


const extendConfig = function(newConfig, origConfig) {
  if (newConfig.exports == null) { newConfig.exports = {}; }
  _.defaults(newConfig, origConfig);
  _.defaults(newConfig.exports, origConfig.exports);
  return newConfig;
};


const makeStandardStore = function(name, storeConfig) {
  extendConfig(storeConfig, new CrudConfig());
  const { actions, store } = makeSimpleStore(storeConfig);
  return { [`${name}Actions`]: actions, [`${name}Store`]: store };
};


export { CrudConfig, makeSimpleStore, extendConfig, makeStandardStore, isNew, STATES };
