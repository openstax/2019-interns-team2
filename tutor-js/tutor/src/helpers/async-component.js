import React from 'react';
import Loadable from 'react-loadable';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import Loading from 'shared/components/loading-animation';

export function asyncComponent(loader, name = '') {
  const loading = ({ error }) => {
    return (
      error ? <ErrorComponent error={error} />
        : <Loading message={`Loading ${name}â€¦`} />
    );
  };
  return Loadable({
    loader, loading,
  });

}

export function loadAsync(resolve, name) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve, name);
}
