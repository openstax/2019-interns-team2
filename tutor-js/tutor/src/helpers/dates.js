import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { reduce } from 'lodash';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}
