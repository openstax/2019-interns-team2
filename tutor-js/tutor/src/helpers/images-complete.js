import { isFunction } from 'lodash';

export default function imagesComplete({
  body = document.body,
  timeoutAfter = 10000, // in ms, 10 seconds
} = {}) {
  return new Promise((resolve, reject) => {
    if (!body || !isFunction(body.querySelectorAll)) {
      resolve([]);
      return;
    }

    const images = Array.from(body.querySelectorAll('img'));
    if (0 === images.length) {
      resolve(images);
      return;
    }
    let complete = 0;
    let pendingTimeout = setTimeout(() => {
      pendingTimeout = null;
      reject(images);
    }, timeoutAfter);
    const markComplete = () => {
      complete += 1;
      if (complete === images.length) {
        if (pendingTimeout) {
          clearInterval(pendingTimeout);
          pendingTimeout = null;
        }
        resolve(images);
      }
    };
    images.forEach((img) => {
      if (img.naturalWidth) {
        markComplete();
      } else {
        img.addEventListener( 'load', markComplete, false );
        img.addEventListener( 'error', markComplete, false );
      }
    });
  });

}
