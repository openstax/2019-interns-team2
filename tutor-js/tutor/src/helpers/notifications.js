import { NotificationActions } from 'shared';
import Router from './router';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
