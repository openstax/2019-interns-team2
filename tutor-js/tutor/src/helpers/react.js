import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { autobind } from 'core-decorators';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType, autobind,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
