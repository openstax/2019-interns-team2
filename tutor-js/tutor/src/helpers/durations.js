import moment from 'moment';
import _ from 'underscore';
import { TimeStore } from '../flux/time';

export default {
  create(startTime, endTime) {
    return moment(startTime).twix(endTime);
  },

  isPastDue({ due_at }) {
    return moment(TimeStore.getNow()).isAfter(due_at);
  },
};
