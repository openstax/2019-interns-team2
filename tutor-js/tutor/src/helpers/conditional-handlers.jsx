import React from 'react';
import { extend } from 'lodash';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const TeacherDashboard = asyncComponent(
  () => import('../screens/teacher-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.currentRole.isTeacher) {
      return <TeacherDashboard {...props} />;
    } else {
      return <StudentDashboard {...props} />;
    }
  };

  return {
    dashboard() { return renderDashboard; },
  };
};

export { getConditionalHandlers };
