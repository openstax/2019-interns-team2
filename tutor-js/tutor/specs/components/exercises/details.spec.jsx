import Details from '../../../src/components/exercises/details';
import { FakeWindow, Factory } from '../../helpers';

jest.mock('../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);

describe('Exercise Details Component', function() {

  let exercises, props, book;

  beforeEach(() => {
    book = Factory.book();
    const pageIds = Array.from(book.pages.byId.keys());
    exercises = Factory.exercisesMap({ book, pageIds, count: 4 });
    props = {
      book,
      exercises,
      pageIds,
      windowImpl:             new FakeWindow,
      selectedExercise:       exercises.array[1],
      onExerciseToggle:       jest.fn(),
      getExerciseIsSelected:  jest.fn().mockReturnValue(false),
      getExerciseActions:     jest.fn().mockReturnValue({}),
      onShowCardViewClick:    jest.fn(),
    };
  });

  it('renders and matches snapshot', () => {
    expect.snapshot(<Details {...props} />).toMatchSnapshot();
  });

  it('can navigate back/forward', () => {
    const deets = mount(<Details {...props} />);
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[1].content.uid}"]`);
    deets.find('.paging-control.next').simulate('click');
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[2].content.uid}"]`);
    deets.find('.paging-control.prev').simulate('click');
    deets.find('.paging-control.prev').simulate('click');
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[0].content.uid}"]`);
    expect(deets).toHaveRendered('.paging-control.prev[disabled]');
  });

});
