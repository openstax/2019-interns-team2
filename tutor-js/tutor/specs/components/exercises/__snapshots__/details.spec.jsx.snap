// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise Details Component renders and matches snapshot 1`] = `
.c0 {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<div
  className="exercise-details"
>
  <div
    className="controls"
  >
    <a
      className="show-cards"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-th fa-w-16 fa-lg ox-icon ox-icon-th c0"
        data-icon="th"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M149.333 56v80c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V56c0-13.255 10.745-24 24-24h101.333c13.255 0 24 10.745 24 24zm181.334 240v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24h101.333c13.256 0 24.001-10.745 24.001-24zm32-240v80c0 13.255 10.745 24 24 24H488c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24H386.667c-13.255 0-24 10.745-24 24zm-32 80V56c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24h101.333c13.256 0 24.001-10.745 24.001-24zm-205.334 56H24c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24zM0 376v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H24c-13.255 0-24 10.745-24 24zm386.667-56H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H386.667c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24zm0 160H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H386.667c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24zM181.333 376v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
       Back to Card View
    </a>
  </div>
  <div
    className="tutor-paging-navigation"
  >
    <a
      aria-controls="paged-content"
      className="paging-control prev"
      disabled={false}
      href="#"
      onClick={[Function]}
      tabIndex={0}
      target="_blank"
      title="Go Back"
    >
      <div
        className="arrow-wrapper"
      >
        <svg
          className="icon arrow left"
          version="1.1"
          viewBox="0 0 370 735"
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <path
            d="M369.225,658.604l0,76.396l-369.225,-369.225l0.3,-0.3l368.925,-365.475l0,77.168l-291.449,290.332l291.449,291.104Z"
          />
        </svg>
      </div>
    </a>
    <div
      aria-live="polite"
      className="paged-content"
      id="paged-content"
      role="region"
      tabIndex="0"
    >
      <div
        className="exercise-content"
      >
        <div
          className="openstax-exercise-preview exercise-card actions-on-side card"
          data-exercise-id="2@1"
          tabIndex={-1}
        >
          <div
            className="card-body"
          >
            <div
              className="exercise-body"
            >
              <div
                className="openstax-question openstax-question-preview"
              >
                <div
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "How could physics be useful in weather prediction?",
                    }
                  }
                />
                <div
                  className="answers-table"
                >
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <div
                        className="correct-incorrect"
                      />
                      <label
                        className="answer-label"
                        htmlFor="2-option-0"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice a:"
                          className="answer-letter"
                          disabled={true}
                        >
                          a
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <div
                        className="correct-incorrect"
                      />
                      <label
                        className="answer-label"
                        htmlFor="2-option-1"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice b:"
                          className="answer-letter"
                          disabled={true}
                        >
                          b
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting how burning fossil fuel releases pollutants.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <div
                        className="correct-incorrect"
                      />
                      <label
                        className="answer-label"
                        htmlFor="2-option-2"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice c:"
                          className="answer-letter"
                          disabled={true}
                        >
                          c
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting how the flowing water affects Earthâ€™s surface.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <div
                        className="correct-incorrect"
                      />
                      <label
                        className="answer-label"
                        htmlFor="2-option-3"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice d:"
                          className="answer-letter"
                          disabled={true}
                        >
                          d
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting the motion of tectonic plates.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <div
                        className="correct-incorrect"
                      />
                      <label
                        className="answer-label"
                        htmlFor="2-option-4"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice e:"
                          className="answer-letter"
                          disabled={true}
                        >
                          e
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                </div>
                <div
                  className="detailed-solution"
                >
                  <div
                    className="header"
                  >
                    Detailed solution
                  </div>
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "four",
                      }
                    }
                  />
                </div>
              </div>
            </div>
            <div
              className="exercise-tags"
            >
              <span
                className="exercise-tag"
              >
                lo:stax-phys:1-2-1
              </span>
              <span
                className="exercise-tag"
              >
                blooms:3
              </span>
              <span
                className="exercise-tag"
              >
                dok:3
              </span>
              <span
                className="exercise-tag"
              >
                time:long
              </span>
              <span
                className="exercise-tag"
              >
                ID:2@1
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <a
      aria-controls="paged-content"
      className="paging-control next"
      disabled={false}
      href="#"
      onClick={[Function]}
      tabIndex={0}
      title="Go Forward"
    >
      <div
        className="arrow-wrapper"
      >
        <svg
          className="icon arrow right"
          version="1.1"
          viewBox="0 0 370 735"
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <path
            d="M0,658.604l0,76.396l369.225,-369.225l-0.3,-0.3l-368.925,-365.475l0,77.168l291.449,290.332l-291.449,291.104Z"
          />
        </svg>
      </div>
    </a>
  </div>
</div>
`;
